pkg/util/cmdutil/console.go

```diff
@@ -93,3 +94,110 @@ func RemoveTralingNewline(s string) string {
 
 	return s
 }
+
+// TableRow is a row in a table we want to print.  It can be a series of a columns, followed
+// by an additional line of information.
+type TableRow struct {
+	Columns        []string
+	AdditionalInfo string
+}
+
+// PrintTable prints a grid of rows and columns.  Width of columns is automatically
+// determined by the max length of the items in each column.
+func PrintTable(table []TableRow) {
+	if len(table) == 0 {
+		return
+	}
+
+	maxColumnWidths := make([]int, len(table[0].Columns))
+	for i := range maxColumnWidths {
+		maxColumnWidths[i] = -1
+	}
+
+	PrintTableEx(table, maxColumnWidths, "  " /*columnGap*/)
+}
+
+const ellipses = "..."
+
+// PrintTableEx prints a grid of rows and columns.  Width of columns is provided.  Column values
+// longer than this width will be automatically truncated.  Use -1 to indicate no max for a
+// particular column.  A gap can be specified between the columns.
+func PrintTableEx(table []TableRow, maxColumnWidths []int, columnGap string) {
+	if len(table) == 0 {
+		return
+	}
+
+	// Figure out the preferred column width for each column.  It will be set to the max length of
+	// any item in that column.  However, it wil lbe clamped to the value in maxColumnWidths if it
+	// is not -1.
+	preferredColumnWidths := make([]int, len(maxColumnWidths))
+
+	for rowIndex, row := range table {
+		columns := row.Columns
+		if len(columns) != len(maxColumnWidths) {
+			panic(fmt.Sprintf(
+				"Error printing table.  Column count of row %v didn't match maxColumnWidths. %v != %v",
+				rowIndex, len(columns), len(maxColumnWidths)))
+		}
+
+		for columnIndex, val := range columns {
+			preferredColumnWidths[columnIndex] = max(preferredColumnWidths[columnIndex], len(val))
+		}
+	}
+
+	for columnIndex := range preferredColumnWidths {
+		maxWidth := maxColumnWidths[columnIndex]
+		if maxWidth != -1 {
+			preferredColumnWidths[columnIndex] = min(preferredColumnWidths[columnIndex], maxWidth)
+		}
+	}
+
+	format := ""
+	for _, maxWidth := range preferredColumnWidths {
+		format += "%-" + strconv.Itoa(maxWidth+len(columnGap)) + "s"
+	}
+	format += "\n"
+
+	columnCount := len(table[0].Columns)
+	columns := make([]interface{}, columnCount)
```

Comment:
Isn't a column always a `string`? If yes, why `[]interface{}`, instead of `[]string`?

Reply:
necessayr for 'splatting' into Printf below.  you can't do  `fmt.Printf(format, columns...)` when `columns` is a `[]string`.  It has to be a `[]interface{}`